### PR TITLE
[AMDGPU] Compiler should synthesize private buffer resource descriptor from flat_scratch_init

### PR DESCRIPTION
**Pull request by alex-t**
This change implements synthesizing the private buffer resource descriptor in the kernel prolog instead of using the preloaded kernel argument.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp [hunk 1]
```diff
@@ -398,6 +398,7 @@ void SIFrameLowering::emitEntryFunctionFlatScratchInit(
 
   Register FlatScrInitLo;
   Register FlatScrInitHi;
+  Register FlatScratchInitReg = AMDGPU::NoRegister;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:401]: Initialization is unnecessary

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp [hunk 2]
```diff
@@ -407,7 +408,7 @@ void SIFrameLowering::emitEntryFunctionFlatScratchInit(
 
     // Find unused reg to load flat scratch init into
     MachineRegisterInfo &MRI = MF.getRegInfo();
-    Register FlatScrInit = AMDGPU::NoRegister;
+    Register FlatScratchInitReg = AMDGPU::NoRegister;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:411]: Initialization is unnecessary

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp [hunk 3]
```diff
@@ -610,11 +612,15 @@ void SIFrameLowering::emitEntryFunctionPrologue(MachineFunction &MF,
...
   MachineRegisterInfo &MRI = MF.getRegInfo();
-  const Function &F = MF.getFunction();
   MachineFrameInfo &FrameInfo = MF.getFrameInfo();
 
   assert(MFI->isEntryFunction());
 
+  bool NeedsFlatScratchInit =
+      MFI->getUserSGPRInfo().hasFlatScratchInit() &&
+      (MRI.isPhysRegUsed(AMDGPU::FLAT_SCR) || FrameInfo.hasCalls() ||
+       (!allStackObjectsAreDead(FrameInfo) && ST.enableFlatScratch()));
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:622]: We could avoid some of this if we fixed #63586

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp [hunk 4]
```diff
@@ -829,11 +834,24 @@ void SIFrameLowering::emitEntryFunctionScratchRsrcRegSetup(
...
-      BuildMI(MBB, I, DL, TII->get(AMDGPU::COPY), ScratchRsrcReg)
-          .addReg(PreloadedScratchRsrcReg, RegState::Kill);
+    if (FlatScratchInit) {
+      I = BuildMI(MBB, I, DL, TII->get(AMDGPU::COPY),
+                  TRI->getSubReg(ScratchRsrcReg, AMDGPU::sub0_sub1))
+              .addReg(FlatScratchInit)
+              .addReg(ScratchRsrcReg, RegState::ImplicitDefine);
+      I = BuildMI(MBB, I, DL, TII->get(AMDGPU::S_MOV_B64),
+                  TRI->getSubReg(ScratchRsrcReg, AMDGPU::sub2_sub3))
+              .addImm(0xf0000000)
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:845]: Avoid hardcoding this with getScratchRsrcWords23()?

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp [hunk 5]
```diff
@@ -829,11 +834,24 @@ void SIFrameLowering::emitEntryFunctionScratchRsrcRegSetup(
...
+              .addReg(ScratchRsrcReg, RegState::ImplicitDefine);
+      return;
+    } else {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:848]: No else after return

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp (hunk 4 above)
Comment (alex-t) [reply llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:845]: From our email conversation:
My question: We have a specialized function that forms the RSRC 48-127 bits SIInstrInfo::getScratchRsrcWords23. Why does not it used for the amdhsa? In other words, do we really always have same constant value in 48-127?

Your answer:  Is a bit historical. Theoretically we were supposed to be able to control a few of the fields in there, but the runtime never implemented it

I went this function in the debugger with "-mtriple=amdgcn-amd-amdhsa" and it returns something different from the 0xf000000. Same time, I have passed PSDB with the 0xf000000 hardcoded.

So, my question is: if it is known that high bits are ignored, could I just call the getScratchRsrcWords23() without revising its logic?

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp (hunk 3 above)
Comment (alex-t) [reply llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:621]: Do you consider committing the current change or want me to fix the https://github.com/llvm/llvm-project/issues/63586 first?

**Comment by alex-t** [comment]
Ping

**Review by slinder1 — COMMENTED** [review]
(Sorry for adding more nits after initial review)

I'm a bit confused about the goal; does this patch just pave the way to actually stop requesting the preloaded RSRC in some cases in the future?

**Review by arsenm — COMMENTED** [review]
We do need a follow up change to prune this from the calling convention lowering. It's going to cause a huge amount of test churn, so it should be done separately

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp [hunk 6]
```diff
@@ -829,6 +833,26 @@ void SIFrameLowering::emitEntryFunctionScratchRsrcRegSetup(
...
+      I = BuildMI(MBB, I, DL, TII->get(AMDGPU::COPY),
+                  TRI->getSubReg(ScratchRsrcReg, AMDGPU::sub0_sub1))
+              .addReg(FlatScratchInit)
+              .addReg(ScratchRsrcReg, RegState::ImplicitDefine);
+      BuildMI(MBB, I, DL, SMovB32, Rsrc2)
+          .addImm(Rsrc23 & 0xffffffff)
+          .addReg(ScratchRsrcReg, RegState::ImplicitDefine);
+
+      BuildMI(MBB, I, DL, SMovB32, Rsrc3)
+          .addImm(Rsrc23 >> 32)
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:851]: Nit: Hi_32

File: llvm/lib/Target/AMDGPU/SIFrameLowering.cpp [hunk 7]
```diff
@@ -829,6 +833,26 @@ void SIFrameLowering::emitEntryFunctionScratchRsrcRegSetup(
...
+      const MCInstrDesc &SMovB32 = TII->get(AMDGPU::S_MOV_B32);
+      Register Rsrc2 = TRI->getSubReg(ScratchRsrcReg, AMDGPU::sub2);
+      Register Rsrc3 = TRI->getSubReg(ScratchRsrcReg, AMDGPU::sub3);
+      uint64_t Rsrc23 = TII->getScratchRsrcWords23();
+      I = BuildMI(MBB, I, DL, TII->get(AMDGPU::COPY),
+                  TRI->getSubReg(ScratchRsrcReg, AMDGPU::sub0_sub1))
+              .addReg(FlatScratchInit)
+              .addReg(ScratchRsrcReg, RegState::ImplicitDefine);
+      BuildMI(MBB, I, DL, SMovB32, Rsrc2)
+          .addImm(Rsrc23 & 0xffffffff)
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:847]: Nit: Lo_32

**Comment by slinder1** [comment]
> > (Sorry for adding more nits after initial review)
> > I'm a bit confused about the goal; does this patch just pave the way to actually stop requesting the preloaded RSRC in some cases in the future?
> 
> Yep. It currently avoids using the preloaded RSRC if the FlatScratchInit is available as RSRC low 64 bits repeat the FlatScratchInit and the hi 64 bits are known const value. However, I still have no idea how to instruct runtime not to set s[0:3] up with the preloaded values in case the whole 128 bits are effectively synthesized in a compile time.

Ah, makes sense; I'm assuming you've seen this, but just in case we do ENABLE_SGPR_PRIVATE_SEGMENT_BUFFER (see https://llvm.org/docs/AMDGPUUsage.html#code-object-v3-kernel-descriptor) which might help. If this simply doesn't work (or isn't always available) then I'm similarly not certain how one achieves this uniformly.

Also, linking to AMDGPUUsage reminded me that we do document the initial SRSRC setup at https://llvm.org/docs/AMDGPUUsage.html#private-segment-buffer . Would you be able to update that as part of the patch? It lives at `llvm/docs/AMDGPUUsage.rst`. At the very least a FIXME/WARNING in there would be good so we don't mislead anyone while we work out the specifics.

**Comment by alex-t** [comment]
> We do need a follow up change to prune this from the calling convention lowering. It's going to cause a huge amount of test churn, so it should be done separately

So, could you please formalize the request a bit and possibly open the Jira ticket?
Just to clarify what exactly did you mean to prune: Rsrc23 constants "black magic" or... what?

**Comment by alex-t** [comment]
> Ah, makes sense; I'm assuming you've seen this, but just in case you haven't we do have ENABLE_SGPR_PRIVATE_SEGMENT_BUFFER (see https://llvm.org/docs/AMDGPUUsage.html#code-object-v3-kernel-descriptor) which might help. If this simply doesn't work (or isn't always available) then I'm similarly not certain how one achieves this uniformly.
> 
I knew we had a bit for this. I mostly was curious how one could set/clear the appropriate bit. Just because I haven't experience with the runtime code and API.

**Comment by alex-t** [comment]
> Also, linking to AMDGPUUsage reminded me that we do document the initial SRSRC setup at https://llvm.org/docs/AMDGPUUsage.html#private-segment-buffer . Would you be able to update that as part of the patch? It lives at `llvm/docs/AMDGPUUsage.rst`. At the very least a FIXME/WARNING in there would be good so we don't mislead anyone while we work out the specifics.

Good Idea. I promise I will do what I can :)

**Comment by slinder1** [comment]
Thank you for the nit changes! This LGTM, as long as you at least put the FIXME in the docs in the initial commit. I'm happy to update it after-the-fact if it unblocks you

**Review by slinder1 — APPROVED** [review]
Thank you, this LGTM!

We can revisit the wording of the whole section when we no longer request the Private Segment Buffer SGPRs unconditionally

File: llvm/docs/AMDGPUUsage.rst [hunk 8]
```diff
@@ -5468,9 +5468,13 @@ If the *Target Properties* column of :ref:`amdgpu-processor-table` specifies
 Instead the flat SCRATCH instructions are used.
 
 Otherwise, Private Segment Buffer SGPR register is used to initialize 4 SGPRs
-that are used as a V# to access scratch. CP uses the value provided by the
-runtime. It is used, together with Scratch Wavefront Offset as an offset, to
-access the private memory space using a segment address. See
+that are used as a V# to access scratch. 
+FIXME: The compiler synthetizes the initialization value for the Private Segment
```
Comment (arsenm) [inline llvm/docs/AMDGPUUsage.rst:5472]: Not sure why this is "FIXME". Also typo 'synthetizes'

**Comment by arsenm** [comment]
> > We do need a follow up change to prune this from the calling convention lowering. It's going to cause a huge amount of test churn, so it should be done separately
> 
> So, could you please formalize the request a bit and possibly open the Jira ticket? Just to clarify what exactly did you mean to prune: Rsrc23 constants "black magic" or... what?

It's really the same ticket. As it stands, this PR is pointless and just increases the instruction count for no benefit. The point was to prune out the resource descriptor in the kernel user SGPRs, which this isn't doing yet. All of the kernel uses of s[0:3] should be removed and shifted down by 4 SGPRs

**Comment by alex-t** [comment]
> As it stands, this PR is pointless and just increases the instruction count for no benefit. The point was to prune out the resource descriptor in the kernel user SGPRs, which this isn't doing yet. All of the kernel uses of s[0:3] should be removed and shifted down by 4 SGPRs

We had a long email conversation regarding this. Let's try one more time.
My question is - how did you intend to get rid of the using s[0:3]?
The buffer_load/store instructions require V# in 4 adjacent SGPRs. How are you supposed to save SGPRs on SRD uses? We could shorten the SGPRs live ranges by rematerializing the SRD before each use but we are not guaranteed to have free SGPRs after the regalloc is done. We could explicitly construct the SRD from the FlatScratrchInit SGPR pair on each use, but we still need two more SGPRs for the high 64  bits.

File: llvm/docs/AMDGPUUsage.rst (hunk 8 above)
Comment (alex-t) [reply llvm/docs/AMDGPUUsage.rst]: removed FIXME and corrected the typo

**Comment by ronlieb** [comment]
@alex-t   could you try applying  your patch to AMD internal gerrit ?
would help us with some down stream merges attn: @saiislam .
thx

**Comment by jplehr** [comment]
I reverted locally and the problems go away.

**Comment by ronlieb** [comment]
our downstream CI is seeing issues with this patch as well , thx for revert

**Comment by alex-t** [comment]
Thanks for catching this. The patch managed to pass PSDB applied to amd-staging. 
So, I thought it was safe to commit.
I will look at the openmp issues closely later today but, at a first glance,  it seems like the private segment buffer gets initialized with the wrong value.

**Comment by alex-t** [comment]
> Hi, this patch may have introduced quite a few breakages in OpenMP GPU Offload buildbots. https://lab.llvm.org/buildbot/#/builders/193/builds/46392 https://lab.llvm.org/staging/#/builders/185/builds/1650 https://lab.llvm.org/staging/#/builders/140/builds/2167
> 
> I'd appreciate if someone can look at this. If you need help with reproducing, configs, etc, please reach out. Happy to help.
 
Could you possibly provide a reproducer case from the build failed?  Or suggest how I can reproduce the failure, preferably w/o cloning/configuring/building the whole project. I would appreciate any help.

**Comment by jplehr** [comment]
Hi Alex, do I understand correctly that the preferred reproducer is intermediate IR + final binary?

**Comment by alex-t** [comment]
> Hi Alex, do I understand correctly that the preferred reproducer is intermediate IR + final binary?

As far as I understand, the error happens in runtime while attempting to read memory beyond the aperture bounds.
So, what I would like ideally is guidance on how to get/configure/build the code that causes an error, or just the assembly itself to inspect the exact value of the V# (SRD) being used. Also, I would need an IR and compiler invocation command line to debug the reasons for setting SRD incorrectly.
Literally, I need to track down the compilation of the code that caused incorrect memory access.
Although, I know almost nothing about the openmp offload project and I would appreciate a reference on the docs that makes me understand better what went on.

**Comment by JonChesterfield** [comment]
> Literally, I need to track down the compilation of the code that caused incorrect memory access. Although, I know almost nothing about the openmp offload project and I would appreciate a reference on the docs that makes me understand better what went on.

It's quite obfuscated by the openmp toolchain providing a user friendly experience. Ultimately there's a single IR blob that gets fed into the backend to build a single code object containing some number of kernels. save-temps might give you that IR, or it might give you some unrelated IR - both have been true at various points in the past.

`clang -fopenmp -###`  should tell you the steps. Some are openmp-specific IR rewriting tools. Somewhere in that pipeline is the IR you seek.

**Pull request merged**